hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/filter/ReclaimableKeyFilter.java

```diff
@@ -0,0 +1,161 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.om.snapshot.filter;
+
+import static org.apache.hadoop.ozone.om.snapshot.SnapshotUtils.isBlockLocationInfoSame;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.ozone.om.KeyManager;
+import org.apache.hadoop.ozone.om.OmSnapshot;
+import org.apache.hadoop.ozone.om.OmSnapshotManager;
+import org.apache.hadoop.ozone.om.OzoneManager;
+import org.apache.hadoop.ozone.om.SnapshotChainManager;
+import org.apache.hadoop.ozone.om.helpers.OmBucketInfo;
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.apache.hadoop.ozone.om.helpers.SnapshotInfo;
+import org.apache.hadoop.ozone.om.lock.IOzoneManagerLock;
+import org.apache.hadoop.ozone.om.snapshot.ReferenceCounted;
+import org.apache.ratis.util.MemoizedCheckedSupplier;
+import org.apache.ratis.util.function.CheckedSupplier;
+
+/**
+ * Filter to return deleted keys which are reclaimable based on their presence in previous snapshot in
+ * the snapshot chain.
+ */
+public class ReclaimableKeyFilter extends ReclaimableFilter<OmKeyInfo> {
+  private final Map<UUID, Long> exclusiveSizeMap;
+  private final Map<UUID, Long> exclusiveReplicatedSizeMap;
+
+  /**
+   * @param currentSnapshotInfo  : If null the deleted keys in AOS needs to be processed, hence the latest snapshot
+   *                             in the snapshot chain corresponding to bucket key needs to be processed.
+   * @param keyManager      : keyManager corresponding to snapshot or AOS.
+   * @param lock                 : Lock for Active OM.
+   */
+  public ReclaimableKeyFilter(OzoneManager ozoneManager,
+                              OmSnapshotManager omSnapshotManager, SnapshotChainManager snapshotChainManager,
+                              SnapshotInfo currentSnapshotInfo, KeyManager keyManager,
+                              IOzoneManagerLock lock) {
+    super(ozoneManager, omSnapshotManager, snapshotChainManager, currentSnapshotInfo, keyManager, lock, 2);
+    this.exclusiveSizeMap = new HashMap<>();
+    this.exclusiveReplicatedSizeMap = new HashMap<>();
+  }
+
+  @Override
+  protected String getVolumeName(Table.KeyValue<String, OmKeyInfo> keyValue) throws IOException {
+    return keyValue.getValue().getVolumeName();
+  }
+
+  @Override
+  protected String getBucketName(Table.KeyValue<String, OmKeyInfo> keyValue) throws IOException {
+    return keyValue.getValue().getBucketName();
+  }
+
+  @Override
```

Comment:
```suggestion
/**
 * Determines whether a deleted key entry is reclaimable by checking its presence in prior snapshots.
 *
 * This method validates the existence of the deleted key in the previous snapshot's key table or file table.
 * If the key is not found in the previous snapshot, it is marked as reclaimable. Otherwise, additional checks 
 * are performed using the previous-to-previous snapshot to confirm its eligibility for reclamation.
 *
 * @param deletedKeyInfo The key-value pair representing the deleted key information.
 * @return {@code true} if the key is reclaimable (not present in prior snapshots), {@code false} otherwise.
 * @throws IOException If an error occurs while accessing snapshot data or key information.
 */
@Override
```